packages/proto-loader/bin/proto-loader-gen-types.ts

```diff
@@ -0,0 +1,745 @@
+    const method = serviceType.methods[methodName];
+    for (const name of [methodName, camelCase(methodName)]) {
+      if (options.includeComments) {
+        formatComment(formatter, method.comment);
+      }
+      const requestType = getTypeInterfaceName(method.resolvedRequestType!);
+      const responseType = getTypeInterfaceName(method.resolvedResponseType!) + '__Output';
+      const callbackType = `(error?: grpc.ServiceError, result?: ${responseType}) => void`;
+      if (method.requestStream) {
+        if (method.responseStream) {
+          // Bidi streaming
+          const callType = `grpc.ClientDuplexStream<${requestType}, ${responseType}>`;
+          formatter.writeLine(`${name}(metadata: grpc.Metadata, options?: grpc.CallOptions): ${callType};`);
+          formatter.writeLine(`${name}(options?: grpc.CallOptions): ${callType};`);
+        } else {
+          // Client streaming
+          const callType = `grpc.ClientWritableStream<${responseType}>`;
+          formatter.writeLine(`${name}(metadata: grpc.Metadata, options: grpc.CallOptions, callback: ${callbackType}): ${callType};`);
+          formatter.writeLine(`${name}(metadata: grpc.Metadata, callback: ${callbackType}): ${callType};`);
+          formatter.writeLine(`${name}(options: grpc.CallOptions, callback: ${callbackType}): ${callType};`);
+          formatter.writeLine(`${name}(callback: ${callbackType}): ${callType};`);
+        }
+      } else {
+        if (method.responseStream) {
+          // Server streaming
+          const callType = `grpc.ClientReadableStream<${responseType}>`;
+          formatter.writeLine(`${name}(argument: ${requestType}, metadata: grpc.Metadata, options?: grpc.CallOptions): ${callType};`);
+          formatter.writeLine(`${name}(argument: ${requestType}, options?: grpc.CallOptions): ${callType};`);
+        } else {
+          // Unary
+          const callType = 'grpc.ClientUnaryCall';
+          formatter.writeLine(`${name}(argument: ${requestType}, metadata: grpc.Metadata, options: grpc.CallOptions, callback: ${callbackType}): ${callType};`);
+          formatter.writeLine(`${name}(argument: ${requestType}, metadata: grpc.Metadata, callback: ${callbackType}): ${callType};`);
+          formatter.writeLine(`${name}(argument: ${requestType}, options: grpc.CallOptions, callback: ${callbackType}): ${callType};`);
+          formatter.writeLine(`${name}(argument: ${requestType}, callback: ${callbackType}): ${callType};`);
+        }
+      }
+    }
+    formatter.writeLine('');
+  }
+  formatter.unindent();
+  formatter.writeLine('}');
+}
+
+function generateServiceHandlerInterface(formatter: TextFormatter, serviceType: Protobuf.Service, options: GeneratorOptions) {
+  if (options.includeComments) {
+    formatComment(formatter, serviceType.comment);
+  }
+  formatter.writeLine(`export interface ${serviceType.name}Handlers extends grpc.UntypedServiceImplementation {`);

```

Comment:
I'm wondering if we could potentially **not** extend from `grpc.UntypedServiceImplementation`?

The problem with using `grpc.UntypedServiceImplementation` is you cannot use this interface for implementation classes. 

What i'd like to do:

```ts
class Server implements ServerHandlers {
  // ... implementation
}
```

But this results in: 

```
Class 'Server' incorrectly implements interface 'ServerHandlers'.
  Index signature is missing in type 'Server'.
```

because `grpc.UntypedServiceImplementation` is defined as:

https://github.com/grpc/grpc-node/blob/21adc97c4ad458b9a09bd83c2cadc7fb2a345367/packages/grpc-js/src/server.ts#L92-L94

The fact that we're now generating types for the server handler, perhaps implies we should not be extending from `grpc.UntypedServiceImplementation`?

Reply:
Hmm, sorry, actually that would not work, because it then would not match the types here:

https://github.com/grpc/grpc-node/blob/21adc97c4ad458b9a09bd83c2cadc7fb2a345367/packages/grpc-js/src/server.ts#L147-L150

`server.addService` requires `UntypedServiceImplementation` type.

Reply:
The problem that solved was doing this:

```js
const handlerImplementation: ServerHandlers = {
  method1: ...
  method2: ...
};
server.addService(serviceDefinition, handlerImplementation);
```

Without extending `UntypedServiceImplementation`, the last line was failing with a similar error to the one you quoted, because `addService` expects an `UntypedServiceImplementation`. If we made the change you suggest without any other changes, I think you would be able to implement that `Server` class, but it would not be accepted by `addService`. I don't know what else to do here, so if you have a suggestion that avoids that problem I'm all ears.

Reply:
Is it worth exploring generic types?

```ts
export declare class Server<T = UntypedServiceImplementation> {
    // ...
    addService(service: ServiceDefinition, implementation: T): void;
    // ...
}
```

```ts
class ExampleServer implements ServerHandlers {
  // ... implementation
}

const server = new grpc.Server<ExampleServer>();
server.addService(proto.chat_package.Chat.service, new ExampleServer());
```

Or we could use generic types on the `addService` method instead.

Reply:
It would be better to have generic types on `addService`. The reason `addService` is a method instead of a constructor argument or something is to allow for adding multiple services to a single server. That wouldn't play nicely with class-level generics.

And now that you mention it, there is the open PR #1561, which is adding a generic type to `addService` to solve a slightly different problem, but it might help with this one too.

Reply:
Thanks for the link to the PR. I see the difficulty now. 

If we add a generic type that defaults to `UntypedServiceImplementation`, we don't enforce any server method type checks. 

```ts
addService<T = UntypedServiceImplementation>(service: ServiceDefinition, implementation: T): void;
```

If we enforce the generic type to be of type `UntypedServiceImplementation`, we can't then implement that type with a class (as typescript will error with "Index signature is missing ").

```ts
addService<T extends UntypedServiceImplementation = UntypedServiceImplementation>(service: ServiceDefinition, implementation: T): void;
```

For the first option, although not ideal, you could argue in a lot of cases the generic type passed to `addService` would be an auto generated type (interface) with method definitions that match the expected method types (`Handler<any, any>`). I still would prefer to have those checks there. I'm not sure how best to achieve this.